handler/src/main/java/io/netty/handler/ssl/ResumptionController.java

```diff
@@ -0,0 +1,172 @@
+    public TrustManager wrapIfNeeded(TrustManager tm) {
+        if (tm instanceof ResumableX509ExtendedTrustManager) {
+            if (PlatformDependent.javaVersion() < 7 || !(tm instanceof X509ExtendedTrustManager)) {
+                throw new IllegalStateException("ResumableX509ExtendedTrustManager implementation must be a " +
+                        "subclass of X509ExtendedTrustManager, found: " + (tm == null ? null : tm.getClass()));
+            }
+            if (!resumableTm.compareAndSet(null, (ResumableX509ExtendedTrustManager) tm)) {
+                throw new IllegalStateException(
+                        "Only one ResumableX509ExtendedTrustManager can be configured for resumed sessions");
+            }
+            return new X509ExtendedWrapTrustManager((X509ExtendedTrustManager) tm, confirmedValidations);
+        }
+        return tm;
+    }
+
+    public void remove(SSLEngine engine) {
+        if (resumableTm.get() != null) {
+            confirmedValidations.remove(unwrapEngine(engine));
+        }
+    }
+
+    public boolean validateResumeIfNeeded(SSLEngine engine)
+            throws CertificateException, SSLPeerUnverifiedException {
+        ResumableX509ExtendedTrustManager tm;
+        boolean valid = engine.getSession().isValid();
+        if (valid && (tm = resumableTm.get()) != null) {
+            Certificate[] peerCertificates = engine.getSession().getPeerCertificates();
+
+            // Unwrap JdkSslEngines because they add their inner JDK SSLEngine objects to the set.
+            engine = unwrapEngine(engine);
+
+            if (!confirmedValidations.remove(engine)) {
+                // This is a resumed session.
+                if (engine.getUseClientMode()) {
+                    // We are the client, resuming a session trusting the server
+                    tm.resumeServerTrusted(chainOf(peerCertificates), engine);
+                } else {
+                    // We are the server, resuming a session trusting the client
+                    tm.resumeClientTrusted(chainOf(peerCertificates), engine);
+                }
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static SSLEngine unwrapEngine(SSLEngine engine) {
+        if (engine instanceof JdkSslEngine) {
+            return ((JdkSslEngine) engine).getWrappedEngine();
+        }
+        return engine;
+    }
+
+    private static X509Certificate[] chainOf(Certificate[] peerCertificates) {
+        if (peerCertificates instanceof X509Certificate[]) {
+            //noinspection SuspiciousArrayCast
+            return (X509Certificate[]) peerCertificates;
+        }
+        X509Certificate[] chain = new X509Certificate[peerCertificates.length];
+        for (int i = 0; i < peerCertificates.length; i++) {
+            Certificate cert = peerCertificates[i];
+            if (cert instanceof X509Certificate) {
+                chain[i] = (X509Certificate) cert;
+            } else {
```

Comment:
This will throw if `cert == null` which it does not for the case when there is a null in it and its of type ` X509Certificate[]`. I think we should be consistent and just ignore a null value and put it in the array.